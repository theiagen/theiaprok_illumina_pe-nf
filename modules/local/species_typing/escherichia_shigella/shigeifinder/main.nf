process SHIGEIFINDER {
    tag "$meta.id"
    label 'process_low'

    container "us-docker.pkg.dev/general-theiagen/staphb/shigeifinder:1.3.5"

    input:
    tuple val(meta), path(assembly)

    output:
    tuple val(meta), path("*_shigeifinder.tsv"), emit: shigeifinder_report
    tuple val(meta), path("shigeifinder_serotype.txt"), emit: shigeifinder_serotype
    tuple val(meta), path("shigeifinder_ipaH_presence_absence.txt"), optional: true, emit: shigeifinder_ipaH_presence_absence
    tuple val(meta), path("shigeifinder_num_virulence_plasmid_genes.txt"), optional: true, emit: shigeifinder_num_virulence_plasmid_genes
    tuple val(meta), path("shigeifinder_cluster.txt"), optional: true, emit: shigeifinder_cluster
    tuple val(meta), path("shigeifinder_O_antigen.txt"), optional: true, emit: shigeifinder_O_antigen
    tuple val(meta), path("shigeifinder_H_antigen.txt"), optional: true, emit: shigeifinder_H_antigen
    tuple val(meta), path("shigeifinder_notes.txt"), optional: true, emit: shigeifinder_notes
    path "versions.yml", emit: versions

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    # capture date
    date | tee DATE

    # ShigEiFinder checks that all dependencies are installed before running
    echo "checking for shigeifinder dependencies and running ShigEiFinder..."
    # run shigeifinder on assembly; default is 4cpus, so turning down to 2 since it's already very fast
    shigeifinder -i ${assembly} \\
        -t ${task.cpus} \\
        --hits \\
        --output ${prefix}_shigeifinder.tsv

    # parse output TSV
    echo "Parsing ShigEiFinder output TSV..."

    # set helpful output strings if field in TSV is blank by overwriting output TXT files
    if [ "\$(head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 2)" == "" ]; then
       echo "ShigEiFinder ipaH field was empty" > shigeifinder_ipaH_presence_absence.txt
    else
       head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 2 > shigeifinder_ipaH_presence_absence.txt
    fi
    if [ "\$(head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 3)" == "" ]; then
       echo "ShigEiFinder number of virulence plasmid genes field was empty" > shigeifinder_num_virulence_plasmid_genes.txt
    else
       head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 3 > shigeifinder_num_virulence_plasmid_genes.txt
    fi
    if [ "\$(head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 4)" == "" ]; then
       echo "ShigEiFinder cluster field was empty" > shigeifinder_cluster.txt
    else
       head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 4 > shigeifinder_cluster.txt
    fi
    if [ "\$(head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 5)" == "" ]; then
       echo "ShigEiFinder serotype field was empty" > shigeifinder_serotype.txt
    else
       head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 5 > shigeifinder_serotype.txt
    fi
    if [ "\$(head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 6)" == "" ]; then
       echo "ShigEiFinder O antigen field was empty" > shigeifinder_O_antigen.txt
    else 
       head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 6 > shigeifinder_O_antigen.txt
    fi
    if [ "\$(head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 7)" == "" ]; then
       echo "ShigEiFinder H antigen field was empty" > shigeifinder_H_antigen.txt
    else
       head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 7 > shigeifinder_H_antigen.txt
    fi
    if [ "\$(head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 8)" == "" ]; then
       echo "ShigEiFinder notes field was empty" > shigeifinder_notes.txt
    else
       head -n 2 ${prefix}_shigeifinder.tsv | tail -n 1 | cut -f 8 > shigeifinder_notes.txt
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        shigeifinder: \$(shigeifinder --version)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    echo "Empty" > "${prefix}_shigeifinder.tsv"
    echo "ShigEiFinder ipaH field was empty" > "shigeifinder_ipaH_presence_absence.txt"
    echo "ShigEiFinder number of virulence plasmid genes field was empty" > "shigeifinder_num_virulence_plasmid_genes.txt"
    echo "ShigEiFinder cluster field was empty" > "shigeifinder_cluster.txt"
    echo "ShigEiFinder serotype field was empty" > "shigeifinder_serotype.txt"
    echo "ShigEiFinder O antigen field was empty" > "shigeifinder_O_antigen.txt"
    echo "ShigEiFinder H antigen field was empty" > "shigeifinder_H_antigen.txt"
    echo "ShigEiFinder notes field was empty" > "shigeifinder_notes.txt"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        shigeifinder: \$(shigeifinder --version)
    END_VERSIONS
    """
}