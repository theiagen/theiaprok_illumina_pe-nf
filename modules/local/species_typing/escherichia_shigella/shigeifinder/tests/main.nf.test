nextflow_process {
    name "Test Process SHIGEIFINDER"
    script "../main.nf"
    process "SHIGEIFINDER"
    
    test("shigeifinder test") {
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    file("$projectDir/bin/M21F02113_contigs.fasta", checkIfExists: true)
                ])
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.shigeifinder_report,
                    process.out.shigeifinder_serotype,
                    process.out.shigeifinder_ipaH_presence_absence,
                    process.out.shigeifinder_num_virulence_plasmid_genes,
                    process.out.shigeifinder_cluster,
                    process.out.shigeifinder_O_antigen,
                    process.out.shigeifinder_H_antigen,
                    process.out.shigeifinder_notes,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("shigeifinder - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    file("$projectDir/bin/473675_contigs.fasta", checkIfExists: true)
                ])
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}