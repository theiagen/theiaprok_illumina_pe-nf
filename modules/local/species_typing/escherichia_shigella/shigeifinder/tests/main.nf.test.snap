{
    "shigeifinder test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_shigeifinder.tsv:md5,2e8e4a466f0cd3712c2ecf7c269e99a9"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "shigeifinder_serotype.txt:md5,4e538ac566c763dfc4105b77c0a2776a"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "shigeifinder_ipaH_presence_absence.txt:md5,c30dba8197a1560649606ea4745e9d04"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "shigeifinder_num_virulence_plasmid_genes.txt:md5,bda81ba88c634b46394ead43aff31ad5"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "shigeifinder_cluster.txt:md5,0d6eb8e4083f85cc42500b2ed76b3eb7"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "shigeifinder_O_antigen.txt:md5,57138138bf0e4e927747b2279c3689c2"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "shigeifinder_H_antigen.txt:md5,08f220a0c52ec2d483b95118b5bee112"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "shigeifinder_notes.txt:md5,d638507c090e6e2516837cf7c7fc81ec"
                ]
            ],
            [
                "versions.yml:md5,2b5cc2dc552ab223bc0f0ec557a622a2"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-06-24T21:10:51.065175839"
    },
    "shigeifinder - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_shigeifinder.tsv:md5,8c0b0472ba6dd8f65926459e9c3f0253"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_serotype.txt:md5,f71779d60e182830ed09056d287a9b85"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_ipaH_presence_absence.txt:md5,01c02267d50ef05d04a758424716498d"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_num_virulence_plasmid_genes.txt:md5,751ad5bbb0c2833f252c9c6c9686274c"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_cluster.txt:md5,a21aa83b0bd35bdc65ed38dae65965c7"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_O_antigen.txt:md5,509da607078f4182648c8b43cd30cb61"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_H_antigen.txt:md5,08f220a0c52ec2d483b95118b5bee112"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_notes.txt:md5,d638507c090e6e2516837cf7c7fc81ec"
                    ]
                ],
                "8": [
                    "versions.yml:md5,2b5cc2dc552ab223bc0f0ec557a622a2"
                ],
                "shigeifinder_H_antigen": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_H_antigen.txt:md5,08f220a0c52ec2d483b95118b5bee112"
                    ]
                ],
                "shigeifinder_O_antigen": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_O_antigen.txt:md5,509da607078f4182648c8b43cd30cb61"
                    ]
                ],
                "shigeifinder_cluster": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_cluster.txt:md5,a21aa83b0bd35bdc65ed38dae65965c7"
                    ]
                ],
                "shigeifinder_ipaH_presence_absence": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_ipaH_presence_absence.txt:md5,01c02267d50ef05d04a758424716498d"
                    ]
                ],
                "shigeifinder_notes": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_notes.txt:md5,d638507c090e6e2516837cf7c7fc81ec"
                    ]
                ],
                "shigeifinder_num_virulence_plasmid_genes": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_num_virulence_plasmid_genes.txt:md5,751ad5bbb0c2833f252c9c6c9686274c"
                    ]
                ],
                "shigeifinder_report": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_shigeifinder.tsv:md5,8c0b0472ba6dd8f65926459e9c3f0253"
                    ]
                ],
                "shigeifinder_serotype": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "shigeifinder_serotype.txt:md5,f71779d60e182830ed09056d287a9b85"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2b5cc2dc552ab223bc0f0ec557a622a2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-06-24T21:10:55.481701971"
    }
}